Java/core/src/main/java/com/amazon/randomcutforest/util/WeightedIndex.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.util;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * a container class that manages weights
+ * 
+ * @param <Q>
+ */
+public class WeightedIndex<Q> {
+
+    public Q index;
+    public float weight;
+
+    public WeightedIndex(Q object, float weight) {
+        this.index = object;
+        this.weight = weight;
+    }
+
+    /**
+     * a generic MonteCarlo sampler that creates an Arraylist of WeightedIndexes
+     * 
+     * @param input       input list of weighted objects
+     * @param seed        random seed for repreoducibility
+     * @param lengthBound a target bound of the length of the list
+     * @param heavyItems  add the items which are 5.0/lengthBound fraction of the
+     *                    weights
+     * @param scale       scale that multiples the weights of the remainder. Note
+     *                    that elements that are sampled are rescaled to have
+     *                    ensured that the total weight (after removal of heavy
+     *                    items) remains the same in expectation
+     * @param <Q>         a generic index type, typically float[] in the current
+     *                    usage
+     * @return a randomly sampled arraylist (which can be the same list) of length
+     *         about LengthBound
+     */
+    public static <Q> List<WeightedIndex<Q>> createSample(List<WeightedIndex<Q>> input, long seed, int lengthBound,
+            boolean heavyItems, double scale) {
+
+        if (input.size() < lengthBound) {
+            return input;
+        }
+
+        ArrayList<WeightedIndex<Q>> samples = new ArrayList<>();
+        Random rng = new Random(seed);
+        double totalWeight = input.stream().map(x -> (double) x.weight).reduce(Double::sum).get();
+        double remainder = totalWeight;
+
+        if (heavyItems) {
+            remainder = input.stream().map(e -> {
+                if (e.weight > totalWeight * 5.0 / lengthBound) {
```

Comment:
5.0 seems arbitrary. What if instead of a `heavyItems` method argument we provide `forceSampleWeightFraction` or something like that. Then the user can say that if a point's weight is greater than a fraction of the total weight that they specify, then it will be included in the sample.

Reply:
Makes sense. The goal was 0.005% ... but setting a fraction directly is better.